### PR TITLE
code, first draft

### PR DESCRIPTION
**Pull request by tien-d**
This PR adds the code for training and evaluating scene landmark detector on indoor6.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -62,14 +62,77 @@ below:
...
+python main.py --action train --dataset_folder /dataset_directory/indoor6/ --scene_id scene1 --output_folder train_log_directory/
+```
+
+* Train using image patches to obtain the scene landmark detector for `scene1`:
+
+```
+python main.py --action train_patches --dataset_folder /dataset_directory/indoor6/ --scene_id scene1 --output_folder train_log_directory/
+```
+
+*** Note that in the previous example codes, all the training output is stored at `train_log_directory/scene1/`. At each epoch, the latest model is kept at `train_log_directory/scene1/model-latest.ckpt`, and valuated on the validation set. The model with the lowest median angular error is stored at `train_log_directory/scene1/model-best_median.ckpt`
```
Comment (snsinha) [inline README.md:105]: codes -> code

Comment (snsinha) [inline README.md:105]: valuated > evaluated

Comment (snsinha) [inline README.md:105]: The model with the lowest median angular error is stored **in**

File: src/main.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument(
+        '--dataset_folder', type=str, required=True,
+        help='Root directory, where all data is stored')
+    parser.add_argument(
+        '--output_folder', type=str, required=True,
+        help='Output folder')
+    parser.add_argument(
+        '--landmark_config', type=str, default='landmarks/landmarks-300',
+        help='Landmark configuration.')
```
Comment (snsinha) [inline src/main.py:17]: help('File containing scene-specific 3D landmarks.');

File: src/main.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    parser.add_argument(
```
Comment (snsinha) [inline src/main.py:18]: help('File containing information about visibility of landmarks in cameras associated with training set.')

File: src/main.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        '--visibility_config', type=str, default='landmarks/visibility_aug-300',
+        help='Visibility configuration.')
+    parser.add_argument(
+        '--scene_id', type=str, default='scene6',
+        help='Scene id')
+    parser.add_argument(
+        '--model', type=str, default='efficientnet',
+        help='Detector model')
```
Comment (snsinha) [inline src/main.py:26]: change 'Detector model' to 'neural net backbone'

File: src/main.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        help='Pretrained detector model')
+    parser.add_argument(
+        '--num_epochs', type=int, default=200,
+        help='number of training epochs')
+    parser.add_argument(
+        '--action', type=str, default='test',
+        help='train/train_patches/test')
+    parser.add_argument(
+        '--training_batch_size', type=int, default=8,
+        help='Batch size using during training')
```
Comment (snsinha) [inline src/main.py:44]: using > used

File: src/utils/pnp.py [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+
+    q = q / np.linalg.norm(q)
+    R = Quaternion2Rotation(q)
+    t = p[:3]
+    P_new = np.hstack([R, t[:, np.newaxis]])
+
+    return P_new
+
+
+def P3PKe(m, X, inlier_thres=1e-5):
```
Comment (snsinha) [inline src/utils/pnp.py:147]: Is this P3P the approach proposed in [Ke and Roumeliotis 2017]. If so, add a comment

**Review by snsinha — APPROVED** [review]
